### ISSUE TITLE
Compiler-cli Error: Compilation failed. Resource file not found: import..... .scss

### ISSUE THREAD

**Issue by xtianus79** [angular/angular#16256]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

**When running the command**

node_modules/.bin/ngc -p tsconfig-aot.json

**the following result occurs:**

`Error: Compilation failed. Resource file not found: /Users/xxxxx/xxxxx/xxxxx/xxxxx/__scss/global
    at ModuleResolutionHostAdapter.readResource (/Users/xxxx/xxxxx/xxx/xxxxx/node_modules/@angular/compiler-cli/src/compiler_host.js:369:19)
    at CompilerHost.loadResource (/Users/xxxxx/xxxxx/xxxx/xxxx/node_modules/@angular/compiler-cli/src/compiler_host.js:243:85)
    at Object.get (/Users/xxxx/xxxx/xxxxx/xxxx/node_modules/@angular/compiler/bundles/compiler.umd.js:24571:107)
    at DirectiveNormalizer._fetch (/Users/mbp06/xxxx/xxxx/xxxxx/node_modules/@angular/compiler/bundles/compiler.umd.js:13206:43)
    at /Users/mbp06/xxx/xxxxx/xxxx/node_modules/@angular/compiler/bundles/compiler.umd.js:13334:53
    at Array.map (native)
    at DirectiveNormalizer._loadMissingExternalStylesheets (/Users/mbp06/xxx/xxxx/xxxx/node_modules/@angular/compiler/bundles/compiler.umd.js:13334:14)
    at /Users/mbp06/xxxx/xxxx/xxxxx/node_modules/@angular/compiler/bundles/compiler.umd.js:13337:26
    at process._tickCallback (internal/process/next_tick.js:109:7)
    at Module.runMain (module.js:607:11)
Compilation failed
`

**Expected behavior**

**Minimal reproduction of the problem with instructions**
the issue is similarly seen here
https://github.com/swimlane/ngx-datatable/issues/459

**What is the motivation / use case for changing the behavior?**
AOT won't compile

**Please tell us about your environment:**

MacOs, VS code, npm

* **Angular version:** 2.0.X
Angular 4.0.3

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

Chrome
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
Node 7.9
TypeScript 2.3.0
Angular 4.0.3

* **Node (for AoT issues):** `node --version` =

**Comment by xtianus79** [comment]
@ericmartinezr thanks for the input.

**Comment by tbosch** [comment]
Hello, we need a reproduction for this issue. Could you create a Github branch for it?
Btw, does the file actually exist when `ngc` runs?

**Comment by xtianus79** [comment]
@tbosch this is always difficult with an enterprise build. What ill try to do is create a shadow build that has the same configs... 

What do you mean does the file actually exist? Yes it does and im sure of this because my scss files use variables that are dependent on this or it would fail.  

The odd thing that i notice is that if i remove every linking to the scss file it will work. ***but i can't comment it out it still recognizes it as the wrong path. I tested this by commenting out and changing the names and it still was finding it but erroring the path was wrong -- which it's not.  

Hope this is helpful.

**Comment by xtianus79** [comment]
Sorry accidentally closed this

**Comment by xtianus79** [comment]
Here is a repo and the command to initial aot

https://github.com/xtianus79/angular-enterprise-template

node_modules/.bin/ngc -p tsconfig-aot.json

lol, i guess the good thing is I have a template of our skeleton so when you ask for a repo it won't be a pain any longer.

**Comment by ewertb** [comment]
Hi Guys

I'm a n00b with this but I got the same problem. I noticed that for some reason everything sitting in a scss directory had the ".scss" extension removed then the file is not found, I added this as a hack:

if (s.indexOf('/scss') > -1) {
			s += '.scss';
		}

and it seemed to find the files. Any clue as to what might be causing that?

Regards

**Comment by xtianus79** [comment]
@tbosch any update on this? I would be super excited if this were a small bug.  Dont' want to hack core files though

**Comment by mlaval** [comment]
I could reproduce the issue thanks to this repo.

The root cause is that Angular only supports `@import` for standard CSS files, but not the extended version of Sass, which is documented here: http://sass-lang.com/documentation/file.SASS_REFERENCE.html#import 

AFAIK, there is no plan to natively support Sass in Angular. So, such transformation should be done as a pre-process before running `ngc`.

@tbosch Can you confirm please?

**Comment by tbosch** [comment]
+1

**Comment by mlaval** [comment]
Closing then

**Comment by xtianus79** [comment]
This is kinda odd.  Saying it is not supported means in other words, Angular won't run a node-sass plugin to properly run the sass files?

@mlaval @tbosch @jelbourn  Doesn't the angular-csi already support sass by giving the option when creating a project. 

this is from the cli repository... 

> Global styles
> The styles.css file allows users to add global styles and supports CSS imports.
> 
> If the project is created with the --style=sass option, this will be a .sass file instead, and the same applies to scss/less/styl.
> 
> You can add more global styles via the apps[0].styles property in angular-cli.json.

This  is from stack 

> When you are creating your project with angular cli try this:
> 
> ng new My_New_Project --style=sass
> 
> This generating all your components with predifined sass files.
> 
> If you want scss syntax create your project with :
> 
> ng new My_New_Project --style=scss
> 
> Cli handles the rest of it.

Why would such a small amount of code to rectify the problem not be the solution instead of saying it is not supported when it clearly is.  

I am asking that you reconsider re-opening this thread and creating a solution that can rectify this issue rather than tabling it.  

The way I am running sass is more or less the way any good sass developer should be utilizing their file structure.  Altering my build process which is intertwined with web pack is not ideal.

**Comment by xtianus79** [comment]
Sorry I couldn't miss this quote.  

> CSS Preprocessor integration
> Angular-CLI supports all major CSS preprocessors:
> 
> sass/scss (http://sass-lang.com/)
> less (http://lesscss.org/)
> stylus (http://stylus-lang.com/)

**Comment by maxailloud** [comment]
Agree with @xtianus79, it feels weird for angular-cli to support preprocessors but not for AoT build.

**Comment by xtianus79** [comment]
@mlaval Would you be able to consider reopening this?  We have a huge enterprise application that we are trying to run in AOT for production and this is the only thing holding us back from AOT.  Sass was used throughout the application.

**Comment by mmathys** [comment]
This is really frustrating, because that very feature of `@import` was available in Angular 2.

As of right now, lots of developers coverting their project from Angular 2 will either encounter this error, forced to create preprocessing build jobs for Sass or smash their whole Sass into one single file. @mlaval does Angular really want that?

**Comment by tbosch** [comment]
@xtianus79  I think there is a confusion: `ngc` (from `@angular/compiler-cli`) is different than `ng` (from `@angular/cli`). `ngc` is a simple command line tool to run AOT, but angular-cli adds a ton of features ontop of this, including sass support.

**Comment by mmathys** [comment]
@tbosch Can we make `ngc` support Sass?

**Comment by tbosch** [comment]
@mmathys Sorry, adding `sass` to `ngc` is currently out of scope, as `ngc` is meant to be a very simple command line utility. Couldn't you just run sass before running `ngc`?

Using `@import` should still work, but sass never worked. Could you confirm this?

**Comment by xtianus79** [comment]
@tbosch running sass prior is not possible because instead of using css we 100% use sass. Importing css files, which there are a few works fine. Just not when a global sass file is imported.  Using global sass times is a very important aspect of sass in general. Much more so than css. That's why this error will most likely be more prevalent as time goes on. 

Another thought on this is when using sass as the file structure webpack handles how this is completed. Is there something that can be done in a webpack setup.  I don't believe so.  The setup is very custom.

Again no consideration can be given to this topic? I again ask to reconsider

**Comment by mmathys** [comment]
@xtianus79 This starter project works with scss: It needs gulp tasks to convert scss to css and do weird cross inject renaming inside source files. _it's ugly but it works._

https://github.com/evertonrobertoauler/cli-universal-demo

**Comment by xtianus79** [comment]
@mmathys using gulp in our build process is not available.

**Comment by xtianus79** [comment]
@tbosch @mlaval 

is the final verdict on this going to be no support for pre-processors such as sass and less?  If so I want to make sure before I pour into an alternative route.  I don't mind waiting for a fix but if there ins't going to be one I guess I will have to find another solution.

**Comment by maxailloud** [comment]
@xtianus79 I discover that angular-cli already supports sass for AoT, it's already working correctly with it. https://github.com/angular/angular-cli/issues/6235

**Comment by mmathys** [comment]
@maxailloud could you elaborate?

**Comment by mmathys** [comment]
@maxailloud does that mean that `ng serve --prod --aot` works?

**Comment by mmathys** [comment]
@maxailloud It doesn't work for me.

**Comment by xetys** [comment]
I tried that right now...it works

**Comment by xetys** [comment]
isn't there any fix to that? It can't be that hard if angular-cli can do that

**Comment by xtianus79** [comment]
@xetys what did you try that works?  in the next comment you ask if there is any fix?

**Comment by xetys** [comment]
I just generated an application using `ng new test --style scss`, added some npm package, containing some SCSS mixins and variables. Then I added an `@import` statement into app.component.scss and used one mixin, and checked proper behavior with `ng serve --aot`.

At [JHipster](https://github.com/jhipster/generator-jhipster) we have this issue currently, as described in https://github.com/jhipster/generator-jhipster/issues/5745 

there we are using webpack, and I am trying to figure out, how to manually apply webpack configuration, to enable using of scss mixins inside those component styles.

**Comment by xtianus79** [comment]
@xetys I am still confused on the comment "i tried that right now... it works..." <<  what works?

**Comment by xetys** [comment]
it works to take a vendor scss mixin/variable for usage inside an included component style.

it doesn't work in JHipster, as we use webpack directly (yet) and trying to use a vendor mixin leads to the "Resource File not found" error

**Comment by xtianus79** [comment]
@xetys @tbosch @mmathys @mlaval Any update on this issue... we are close to going to production and would love to put this sweet candy up on top.  Make the application go hummmmm.

**Comment by xetys** [comment]
seems to be still present, as it was reported today again at JHipsters issue tracker

**Comment by tbosch** [comment]
Sorry, but the compiler-cli package is supposed to be the minimal angular compiler. For features like scss support you need a build process that first converts these into regular CSS files.

**Comment by tbosch** [comment]
Eg just create a shell script that first runs scss and then NGC.

**Comment by xtianus79** [comment]
@tbosch again you're asking to remove the workflow from webpack and the workflow process in general.  

As of now, a user can simply create a component... a corresponding scss file and continue working like nothing has happened.  

We have in our enterprise build as of this moment.... over 74 main components... not every single one has a corresponding scss file but most of them do. 

Again, I would beg for sass and less to be treated like first class citizens here.  I can't imagine why not.  if you are going to put something in this would be it.  How much effort does this really add?

**Comment by xtianus79** [comment]
Our first Angular 2 app launched in production and we were not able to utilize the AOT configuration.  This is going to be my last effort on this subject to ask you wonderful people to reconsider this decision.  

@tbosch @IgorMinar @jelbourn @mmathys @mary-poppins @DzmitryShylovich 

If you support css and support using a file such as scss why would you not support the basic functionality of what that file does.  All we are asking here is for the file to do what it does as natively.  Which by the way is the same thing a css file does. @import.  

Now, I am understanding of the consideration of not supporting it but really what is the harm or reason for not supporting it?  Is it difficult or to complex to code for?  I don't think that is it.  Is it something else to maintain?  Perhaps, but this is an open source project.  I am sure there are scores of people that would want this functionality that would be happy to maintain this segment of code.  

This type of configuration exists in the CLI but not in a native webpack configuration not using the cli.  That seems very counterproductive and a limiting reagent to forcing everyone to use angular-cli.  Now I have no issue using angular cli but like I have said before it is to boxy in configuration and limiting... perhaps that has changed and configuration is more open as I have not tried it out in some time.  

Also, where is the miscommunication coming from.  Either you guys aren't understanding our build process and why this is important or we are not understanding your proposal for an alternative.  

Let's be fair here, if there is an alternative let's propose it and work through what that build process would look like...  If it is fair and feasible let's document it and let it exist for those who need to utilize it.  If it is not fair and feasible can you all then consider adding this functionality in.  

So in the end I propose we work together and get this issue resolved and not just let it linger.  I am more than willing to do anything to help and document the final results but this issue is a little frustrating and I want to tackle it.  

Thanks for your consideration

**Comment by tbosch** [comment]
I think there is a bit of a misunderstanding here:
`ngc` is a command line tool. Therefore my proposal was in terms of command line build processes.

As you are using webpack, how are you calling into `ngc`? Are you using the Angular webpack plugin (`@ngtools/webpack`)? Or do you have a custom built webpack plugin?

Pinging @hansl from the angular-cli team, as he knows way more about webpack than I do...

**Comment by hansl** [comment]
Hi guys,

I agree with @tbosch, there's a lot of misunderstanding about the different tools that the Angular team provide (and the CLI subteam):

* `ngc` is a tool that compiles your code, but does not provide any kind of plugin or parser. This is perfect when you want to pre-compile your SASS files or use a simpler pipeline where each tools work on hermetic directories (bazel and broccoli come to mind). `ngc` does NOT support compiling SASS, and will read files from `styleUrls` as CSS, even if they don't have the CSS extension.
* `angular-cli` is a tool to generate and manage your project. It's much more opinionated and hand holding, but in the end is the easiest to use and maintain.
* `@ngtools/webpack` is the webpack plugin that is distributed and maintained by the CLI team for webpack. It is made available for people who want to use the powerful webpack build configuration from the CLI without being walled into the way the CLI itself handles things.

What you want is probably using the `@ngtools/webpack` webpack plugin and loader in your webpack config, and it will support handling SCSS just as if you were compiling it manually. You can checkout the documentation (partial) at https://www.npmjs.com/package/@ngtools/webpack

The release schedule for the webpack plugin is the same as the CLI.

**Comment by hansl** [comment]
Please note that we're still working on our next generation CLI, which you **will** be able to use on non-CLI projects if you only need, e.g., the build system, and which you will be able to plug into and modify the behaviour to fit your project (instead of the other way around).

**Comment by StephenFluin** [comment]
@xtianus79 @mmathys 

I typed this up before Hans had posted, so it repeats a lot of what he says, but hopefully it clarifies the same concept.

"Angular" the platform supports developers using SASS/SCSS/LESS or just pure CSS when writing applications. We support this at the component  level, or global styles.  This should include any `import` commands. We also support this in AOT mode in the same way.

`ngc` or the Angular Compiler operates at a lower level depending entirely on browser primitives to apply styles. `ngc` is the tool we offer (and use as part of the Angular CLI or `ng`) to perform Ahead of Time compilation of components, modules, etc.

The way we offer SASS/SCSS/LESS support in the CLI is by adding the style compilation to the build pipeline, PRIOR to invoking `ngc`. This means when you run `ng build -prod` we run the style preprocessor, then collect metadata, then compile, then minify and bundle, giving you nice clean ES5 output that can be run in the browser.

If you want to run `ngc` yourself, you will need to compile styles (just like we do with the CLI).

Am I misunderstanding the use case?

**Comment by xtianus79** [comment]
@StephenFluin As of now there are 2 sets of SCSS files that are compiled 2 different ways.  I am pretty sure the main styles.scss is compiled in the manner in which you are saying... (perhaps I have them reversed) And the other component.scss sheets that go with our components are the ones I am referring to that are styled during "saving" per the webpack configuration.  

The confusion lies in, how do you configure the style sheets to act the way you want during development and opposingly have them pre-built during prod  / qa builds.  That is perhaps the part that is escaping me at the moment.

**Comment by MohibWasay** [comment]
So has anyone able to reach the conclusion. Mr Eiffel is curious.

**Comment by xtianus79** [comment]
@StephenFluin @tbosch @MohibWasay @mmathys @xetys @maxailloud and others.... I can officially say the solution of using @ngtools/webpack worked.  

Here is a working repo of the changes I made mirroring for the most part our production skeleton build.  
[https://github.com/xtianus79/angular-enterprise-template](
https://github.com/xtianus79/angular-enterprise-template)

I will say this this.  Grab your bootstraps because if you thought TypeScript was strict before well that was just your .ts files.  Since AOT is going straight to the source it will typecheck the templated files too.... so bear with it.  

It will not produce the dist folder section until your error free.  That is definately a difference I noticed in version 1.5 as opposed to the ejected version that angular-cli was using of version 1.1 I believe.  

***Sidenote I really need to find a good template linter now that is AOT compatible.  Any suggestions?  

So in all, my suggestion is to use the build you are starting with via webpack and start there.  The only thing I really had to do here is intercept the ts loader configuration.  Everything else was pretty much the same.  

Again, I used the latest version I could find which was @ngtools/webpack 1.5  

I think that is probably / perhaps the most confusing thing about the process is configuring which versions of webpack, TS, and @ngtools/webpack to be utilizing together to get an expected result.  But it does work and this process does exactly what was described before by Tbosch and Stephen... lets your webpack run and build then runs the AOT of the angular application.  

so my main suggestion here would be provide versioning for the @ngtools!

After doing some performance benchmarking and testing I will post the results.  

Thanks to the Angular team for being super helpful and supportive in this issue I hope it can help others.
